File: zxing/src/main/java/com/mazaiting/zxing/CaptureActivity.java
package com.mazaiting.zxing;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.AssetFileDescriptor;
import android.graphics.Bitmap;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Vibrator;
import android.view.SurfaceHolder;
import android.view.SurfaceHolder.Callback;
import android.view.SurfaceView;
import android.widget.Toast;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.Result;
import com.mazaiting.zxing.camera.CameraManager;
import com.mazaiting.zxing.util.InactivityTimer;
import com.mazaiting.zxing.view.ViewfinderView;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

/**
 * 扫描二维码界面
 */
public class CaptureActivity  extends AppCompatActivity implements Callback{
    public static final int PERMISSION_CODE = 0x10000;
    /** 扫描结果键值 */
    public static final String SCAN_RESULT = "result";
    private CaptureActivityHandler handler;
    //扫描框view
    private ViewfinderView viewfinderView;
    private boolean hasSurface;
    private Vector<BarcodeFormat> decodeFormats;
    private String characterSet;
    private InactivityTimer inactivityTimer;
    private MediaPlayer mediaPlayer;
    private boolean playBeep;
    private static final float BEEP_VOLUME = 0.10f;
    private boolean vibrate;
//    private Button cancelScanButton;

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_capture);
        
        requestPermission();
        
        CameraManager.init(getApplication());
        viewfinderView = this.findViewById(R.id.zxing_viewfinder_view);
//        cancelScanButton = this.findViewById(R.id.zxing_btn_cancel_scan);
        hasSurface = false;
        inactivityTimer = new InactivityTimer(this);
    }
    
    /**
     * 请求必要权限
     */
    private void requestPermission() {
        // 判断版本, 6.0以上请求必要权限
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissionForStorageAndCamera();
        }
    }
    
    /**
     * 请求存储权限与照相机权限
     */
    @RequiresApi(api = Build.VERSION_CODES.M)
    private void requestPermissionForStorageAndCamera() {
        List<String> list = new ArrayList<>();
        // 检测是否具有写入内存卡权限
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            list.add(Manifest.permission.WRITE_EXTERNAL_STORAGE);
        }
        // 检测是否具有使用照相机权限
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            list.add(Manifest.permission.CAMERA);
        }
        if (list.size() > 0) {
            String[] strings = new String[list.size()];
            list.toArray(strings);
            // 请求权限
            requestPermissions(strings, PERMISSION_CODE);
        }
    }
    
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (PERMISSION_CODE == requestCode) {
        
        }
    }
    
    @Override
    protected void onResume() {
        super.onResume();
        SurfaceView surfaceView = findViewById(R.id.zxing_preview_view);
        SurfaceHolder surfaceHolder = surfaceView.getHolder();
        if (hasSurface) {
            initCamera(surfaceHolder);
        } else {
            surfaceHolder.addCallback(this);
            surfaceHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS);
        }
        decodeFormats = null;
        characterSet = null;

        playBeep = true;
        AudioManager audioService = (AudioManager) getSystemService(AUDIO_SERVICE);
        if (audioService.getRingerMode() != AudioManager.RINGER_MODE_NORMAL) {
            playBeep = false;
        }
        initBeepSound();
        vibrate = true;

        //quit the scan view
//        cancelScanButton.setOnClickListener(new View.OnClickListener() {
//
//            @Override
//            public void onClick(View v) {
//                CaptureActivity.this.finish();
//            }
//        });
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (handler != null) {
            handler.quitSynchronously();
            handler = null;
        }
        CameraManager.get().closeDriver();
    }

    @Override
    protected void onDestroy() {
        inactivityTimer.shutdown();
        super.onDestroy();
    }

    /**
     * Handler scan result
     * @param result
     * @param barcode
     */
    public void handleDecode(Result result, Bitmap barcode) {
        inactivityTimer.onActivity();
        playBeepSoundAndVibrate();
        String resultString = result.getText();
        //FIXME
        if (resultString.equals("")) {
            Toast.makeText(CaptureActivity.this, "Scan failed!", Toast.LENGTH_SHORT).show();
        }else {
            Intent resultIntent = new Intent();
            Bundle bundle = new Bundle();
            bundle.putString("result", resultString);
            resultIntent.putExtras(bundle);
            this.setResult(RESULT_OK, resultIntent);
        }
        CaptureActivity.this.finish();
    }

    private void initCamera(SurfaceHolder surfaceHolder) {
        try {
            CameraManager.get().openDriver(surfaceHolder);
        } catch (IOException | RuntimeException ioe) {
            return;
        }
        if (handler == null) {
            handler = new CaptureActivityHandler(this, decodeFormats,
                    characterSet);
        }
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width,
                               int height) {

    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        if (!hasSurface) {
            hasSurface = true;
//            initCamera(holder);
        }

    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        hasSurface = false;

    }

    public ViewfinderView getViewfinderView() {
        return viewfinderView;
    }

    public Handler getHandler() {
        return handler;
    }

    public void drawViewfinder() {
        viewfinderView.drawViewfinder();

    }

    //初始化“哔”声音文件
    private void initBeepSound() {
        if (playBeep && mediaPlayer == null) {
            // The volume on STREAM_SYSTEM is not adjustable, and users found it
            // too loud,
            // so we now play on the music stream.
            setVolumeControlStream(AudioManager.STREAM_MUSIC);
            mediaPlayer = new MediaPlayer();
            mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
            mediaPlayer.setOnCompletionListener(beepListener);
            //打开文件
            AssetFileDescriptor file = getResources().openRawResourceFd(
                    R.raw.beep);
            try {
                mediaPlayer.setDataSource(file.getFileDescriptor(),
                        file.getStartOffset(), file.getLength());
                file.close();
                mediaPlayer.setVolume(BEEP_VOLUME, BEEP_VOLUME);
                mediaPlayer.prepare();
            } catch (IOException e) {
                mediaPlayer = null;
            }
        }
    }

    private static final long VIBRATE_DURATION = 200L;

    //播放声音
    private void playBeepSoundAndVibrate() {
        if (playBeep && mediaPlayer != null) {
            mediaPlayer.start();
        }
        if (vibrate) {
            Vibrator vibrator = (Vibrator) getSystemService(VIBRATOR_SERVICE);
            vibrator.vibrate(VIBRATE_DURATION);
        }
    }

    /**
     * When the beep has finished playing, rewind to queue up another one.
     * 当声音结束播放的时候，倒回以将下一个加入队列
     */
    private final MediaPlayer.OnCompletionListener beepListener = new MediaPlayer.OnCompletionListener() {
        public void onCompletion(MediaPlayer mediaPlayer) {
            mediaPlayer.seekTo(0);
        }
    };
}

File: zxing-ktx/src/main/java/com/mazaiting/zxing/CaptureActivity.kt
package com.mazaiting.zxing

import android.Manifest
import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.media.AudioAttributes
import android.media.AudioManager
import android.media.MediaPlayer
import android.os.*
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.zxing.Result
import com.mazaiting.log.L
import com.mazaiting.permission.PermissionResult
import com.mazaiting.permission.Permissions
import com.mazaiting.permission.util.PermissionSettingUtil
import com.mazaiting.permission.util.PermissionUtil
import com.mazaiting.permission.util.State
import com.mazaiting.zxing.camera.CameraManager
import com.mazaiting.zxing.util.InactivityTimer
import com.mazaiting.zxing.view.ViewfinderView
import java.io.IOException

/**
 * 扫描二维码界面
 * @use 使用方法
 * ```
 *  /**
     * 二维码扫描
     */
    private fun qrCodeScan() {
      val openCameraIntent = Intent(this@MainActivity, CaptureActivity::class.java)
      startActivityForResult(openCameraIntent, QR_CODE_QUERY)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
      super.onActivityResult(requestCode, resultCode, data)
      if (resultCode == Activity.RESULT_OK) {
        when (requestCode) {
          QR_CODE_QUERY -> {
            val result = data!!.getStringExtra("result")
            Toast.makeText(this, result, Toast.LENGTH_SHORT).show()
          }
          else -> {}
        }
      }
    }
 * ```
 */
@Permissions([Manifest.permission.CAMERA, Manifest.permission.WRITE_EXTERNAL_STORAGE], CaptureActivity.SCAN_PERMISSION_CODE)
class CaptureActivity : AppCompatActivity(), SurfaceHolder.Callback {
  
  companion object {
    /** 请求权限代码 */
    const val SCAN_PERMISSION_CODE = 0x1000
    /** 蜂鸣声时长 */
    private const val SCAN_BEEP_VOLUME = 0.10f
    /** 震动时长 */
    private const val SCAN_VIBRATE_DURATION = 200L
    /** 扫描结果键值 */
    const val SCAN_RESULT_TEXT = "result"
    /** 扫描结果图像 */
    const val SCAN_RESULT_BITMAP = "qr_code"
  }
  
  /** 扫描界面的Handler */
  private var mHandler: CaptureActivityHandler? = null
  /** 扫描框view */
  private var mViewfinderView: ViewfinderView? = null
  /** Surface是否可用 */
  private var isHasSurface: Boolean = false
  /** 计时器 */
  private var mInactivityTimer: InactivityTimer? = null
  /** 音频播放器 */
  private var mMediaPlayer: MediaPlayer? = null
  /** 是否播放声音 */
  private var isPlayBeep: Boolean = false
  /** 是否震动 */
  private var isVibrate: Boolean = false
  /** Surface控制器 */
  private var mSurfaceHolder: SurfaceHolder? = null
  
  /**
   * 播放监听, 播放完成后, 将进度置为0
   */
  private val mBeepListener = MediaPlayer.OnCompletionListener {
    // 将播放进度置0
    mediaPlayer -> mediaPlayer.seekTo(0)
  }
  
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    // 保持屏幕常亮
    window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
//    initDebug()
//    // 请求权限
//    requestPermission()
    setContentView(R.layout.activity_capture)
    initDebug()
    // 请求权限
//    PermissionUtil.requestPermission(this)
//    requestPermission()
    // 相机管理者初始化
    CameraManager.init(application)
    // 初始化画中画
    mViewfinderView = this.findViewById(R.id.zxing_viewfinder_view)
    // 无Surface
    isHasSurface = false
    // 初始化定时器
    mInactivityTimer = InactivityTimer(this)
  }
  
  /**
   * 初始化Debug
   */
  private fun initDebug() {
    if (BuildConfig.DEBUG) {
      L.setProp(BuildConfig.DEBUG, "SCAN")
    }
  }
  
  override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
//    if (SCAN_PERMISSION_CODE == requestCode) {
//      // TODO 此时如果有拒绝的权限，则关闭当前界面
//      L.d("onRequestPermissionResult")
//    }
    PermissionUtil.onRequestPermissionsResult(this, requestCode, permissions, grantResults)
  }
  
  /**
   * 权限请求结果处理
   * @param state 权限授权状态
   * @param permissions 当状态为Succeed时权限列表为空
   */
  @PermissionResult(SCAN_PERMISSION_CODE)
  fun permissionResult(state: State, permissions: List<String>?) {
    when (state) {
      State.DENIED -> // 再次请求权限
        PermissionUtil.requestPermission(this)
      State.SUCCESS -> // Toast.makeText(this, "权限申请成功!", Toast.LENGTH_SHORT).show()
        initCamera()
      State.NOT_SHOW -> notShow(permissions!!)
    }
  }
  
  /**
   * 点击不再提示后并拒绝
   */
  private fun notShow(permissions: List<String>) {
    val sb = StringBuilder()
    // 迭代读取权限
    permissions.forEach { permission ->
      sb.append("$permission\n")
    }
    // 跳转到应用设置界面
    AlertDialog.Builder(this)
            .setTitle("友情提示！")
            .setMessage("请跳转到设置界面同意下面权限：\n" + sb.toString())
            .setPositiveButton("同意") { _, _ ->
              // 跳转到应用设置界面
//              goToAppSetting(this)
              PermissionSettingUtil.gotoPermissionActivity(this)
            }
            .setNegativeButton("关闭") { _, _ ->
              // 关闭当前页面
              finish()
            }
            .setCancelable(false).show()
  }
  
  override fun onResume() {
    super.onResume()
    // 获取SurfaceView
    val surfaceView = findViewById<SurfaceView>(R.id.zxing_preview_view)
    // 获取SurfaceHolder
    val surfaceHolder = surfaceView.holder
    // 如果已经有Surface, 则进行相机初始化
    if (isHasSurface) {
      mSurfaceHolder = surfaceHolder
      // 初始化相机
      initCameraWithPermission()
    } else {
      // 添加回调
      surfaceHolder.addCallback(this)
      // 设置类型
//      mSurfaceHolder.setType(SurfaceHolder.SURFACE_TYPE_PUSH_BUFFERS)
    }
    // 设置允许蜂鸣
    isPlayBeep = true
    // 获取声音服务
    val audioService = getSystemService(Context.AUDIO_SERVICE) as AudioManager
    // 判断蜂鸣模式
    if (audioService.ringerMode != AudioManager.RINGER_MODE_NORMAL) {
      // 如果不为normal则不播放
      isPlayBeep = false
    }
    // 初始化蜂鸣声音
    initBeepSound()
    // 设置允许震动
    isVibrate = true
  }
  
  override fun onPause() {
    super.onPause()
    // 清除Handler
    if (mHandler != null) {
      // 同步退出
      mHandler!!.quitSynchronously()
      mHandler = null
    }
    // 关闭相机设备
    CameraManager.get().closeDriver()
  }
  
  override fun onDestroy() {
    // 关闭定时器
    mInactivityTimer!!.shutdown()
    super.onDestroy()
  }
  
  /**
   * 处理返回结果
   * @param result 扫描结果
   */
//  fun handleDecode(result: Result) {
  fun handleDecode(result: Result, bitmap: Bitmap) {
    // Activity处理结果
    mInactivityTimer!!.onActivity()
    // 播放声音和震动
    playBeepSoundAndVibrate()
    // 获取结果
    val resultString = result.text
    //FIXME
    if (resultString == "") {
      // 如果内容为空, 提示扫描失败
      Toast.makeText(this@CaptureActivity, resources.getString(R.string.scan_failed), Toast.LENGTH_SHORT).show()
    } else {
      // 设置数据
      val resultIntent = Intent()
      val bundle = Bundle()
      // 存储内容
      bundle.putString(SCAN_RESULT_TEXT, resultString)
      //TODO 待确认是否需要存储bitmap
      // 存储图片
//      bundle.putParcelable(SCAN_RESULT_BITMAP, bitmap)
      resultIntent.putExtras(bundle)
      // 设置扫描结果返回
      this.setResult(AppCompatActivity.RESULT_OK, resultIntent)
    }
    // 关闭当前页面
    this@CaptureActivity.finish()
  }
  
  /**
   * 初始化相机
   */
  private fun initCameraWithPermission() {
    PermissionUtil.requestPermission(this)
  }
  
  /**
   * 初始化相机
   */
  private fun initCamera() {
    try {
      // 打开设备
      mSurfaceHolder?.let { CameraManager.get().openDriver(it) }
    } catch (ioe: IOException) {
      return
    }
    // 如果Handler为空, 则初始化
    if (mHandler == null) {
      // 初始化Handler对象
      mHandler = CaptureActivityHandler(this)
    }
  }
  
  override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {}
  
  override fun surfaceCreated(holder: SurfaceHolder) {
    // 判断Surface是否可用
    if (!isHasSurface) {
      isHasSurface = true
      mSurfaceHolder = holder
      // 初始化相机
      initCameraWithPermission()
    }
  }
  
  override fun surfaceDestroyed(holder: SurfaceHolder) {
    // surface销毁的时候, 将surface置为不可用
    isHasSurface = false
  }
  
  /** 获取ViewfinderView对象 */
  fun getViewfinderView(): ViewfinderView = mViewfinderView!!
  
  /** 返回Handler */
  fun getHandler(): Handler? = mHandler
  
  /** 获取画中画内容 */
  fun drawViewfinder() = mViewfinderView!!.drawViewfinder()
  
  /**
   * 初始化蜂鸣声音文件
   */
  private fun initBeepSound() {
    // 判断是否允许蜂鸣并且播放声音对象为空
    if (isPlayBeep && mMediaPlayer == null) {
      // 设置声音控制流为music, 此处为当前Activity中的
      volumeControlStream = AudioManager.STREAM_MUSIC
      // 初始化播放器
      mMediaPlayer = MediaPlayer()
      // 设置声音流类型
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
        mMediaPlayer!!.setAudioAttributes(AudioAttributes.Builder().setLegacyStreamType(AudioManager.STREAM_MUSIC).build())
      } else {
        mMediaPlayer!!.setAudioStreamType(AudioManager.STREAM_MUSIC)
      }
      // 设置播放完成监听
      mMediaPlayer!!.setOnCompletionListener(mBeepListener)
      //打开文件
      val file = resources.openRawResourceFd(R.raw.beep)
      try {
        // 设置声音源你
        mMediaPlayer!!.setDataSource(file.fileDescriptor,
                file.startOffset, file.length)
        // 关闭文件
        file.close()
        // 设置音量
        mMediaPlayer!!.setVolume(SCAN_BEEP_VOLUME, SCAN_BEEP_VOLUME)
        // 准备播放
        mMediaPlayer!!.prepare()
      } catch (e: IOException) {
        mMediaPlayer = null
      }
    }
  }
  
  /**
   * 播放声音和震动
   */
  private fun playBeepSoundAndVibrate() {
    // 判断是否播放蜂鸣
    if (isPlayBeep && mMediaPlayer != null) {
      mMediaPlayer!!.start()
    }
    // 判断是否震动
    if (isVibrate) {
      // 获取震动服务
      val vibrator = getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
      // 震动
      if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O) {
        vibrator.vibrate(VibrationEffect.createOneShot(SCAN_VIBRATE_DURATION, VibrationEffect.DEFAULT_AMPLITUDE))
      } else {
        vibrator.vibrate(SCAN_VIBRATE_DURATION)
      }
    }
  }
  
}
